### PR TITLE
feat: add editor background image

### LINKED ISSUE #16: Optionally display the project label as a background image of the whole window
When working with multiple open windows in IntelliJ, it is often hard to tell which window is which, e.g. when switching windows using mission control because the label in the status bar is not easily readable. 

When discussing with my colleagues at work, several of them mentioned it would be very helpful, if one could optionally have the project label in a much bigger way as a background image of the whole editor window. 

Since I would love to have that too, I've prototyped a bit and am fairly confident, that this would improve the UX in such cases.

@drinchev Would you be OK to including that in the project-label plugin?
Then I'll go ahead and polish things up to create a PR.

My current implementation looks like this:
 
SCR-20230407-ithd
SCR-20230407-itjy

### PR DESCRIPTION
**Pull request by simschla**
As discussed in #16 here the PR to discuss further.

About the application- vs. project-level-preferences: I think it might be a good idea to also let the user enable the background image on the application level; however, I'm uncertain if we can easily fit that into the current preferences UI in a way that is still easy to understand for the end users. Maybe we need to separate the Application Preferences from the Project Preferences into two preference panels? I have, however, not found a similar case in my IntelliJ setup or any of the plugins I've installed.

I'd love it if you'd try the current solution and perhaps get a feel of what would be a good way forward.

This PR resolves #16.

### CODE REVIEW

File: src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                final int paddingY = JBUIScale.scale(100);
+                // we want to have a 800x600 image, but prefer it to have the same ratio as the project label, so we scale it accordingly
+                final int targetWidth = 800;
+                final int targetHeight = (int) Math.round(targetWidth / preferredImageRation.getWidth() * preferredImageRation.getHeight());
+                bufferedImage = renderer.renderLabel(
+                        new Dimension(targetWidth+2*paddingX, targetHeight+2*paddingY),
+                        new Dimension(paddingX, paddingY));
+
+                Path filePath = Files.createTempFile("project-label", ".png");
+                filePath.toFile().deleteOnExit();
```
Comment (simschla) [inline src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java]: Currently, I delete the file on JVM exit and regenerate a new one when opening the IDE again. Alternatively, we could store the image in a more persistent space and reuse it. Question there would be: where to store it? Or maybe the temp-file approach is good enough? 🤔

File: src/main/java/com/drinchev/projectlabel/resources/ui/PluginConfiguration.java [hunk 2]
```diff
@@ -24,13 +29,32 @@ public class PluginConfiguration {
...
 
+    private static final String BACKGROUND_IMAGE_POSITION_TOP_LEFT = "Top Left";
+    private static final String BACKGROUND_IMAGE_POSITION_TOP_RIGHT = "Top Right";
+    private static final String BACKGROUND_IMAGE_POSITION_BOTTOM_RIGHT = "Bottom Right";
+    private static final String BACKGROUND_IMAGE_POSITION_BOTTOM_LEFT = "Bottom Left";
+    private static final String BACKGROUND_IMAGE_POSITION_CENTER = "Center";
+
+    private final static Map<String, Icon> LABEL_POSITIONS = Collections.unmodifiableMap(new LinkedHashMap<>() {{ // linked hash map to preserve order
+        put(BACKGROUND_IMAGE_POSITION_CENTER, AllIcons.General.FitContent);
+        put(BACKGROUND_IMAGE_POSITION_TOP_LEFT, AllIcons.Actions.MoveToTopLeft);
```
Comment (drinchev) [inline src/main/java/com/drinchev/projectlabel/resources/ui/PluginConfiguration.java]: For some wierd reason those icons are not loaded properly. 

image

File: src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    private boolean shouldShowBackgroundImage() {
+        return preferences.backgroundImagePosition() != BackgroundImagePosition.HIDDEN;
+    }
+
+    private void setImageToIDE() {
+        PropertiesComponent prop = projectLevelPropertiesComponent();
+        String opacity = String.valueOf(preferences.backgroundImageOpacity());
+        String imageProp = String.format("%s,%s,plain,%s", resultingImage, opacity, preferences.backgroundImagePosition().name().toLowerCase());
+        String prev_editor = prop.getValue(IdeBackgroundUtil.EDITOR_PROP);
+        prop.setValue(IdeBackgroundUtil.EDITOR_PROP, imageProp); // statt plain: tile (wiederholen), scale (zoomen)
```
Comment (drinchev) [inline src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java]:
```suggestion
        prop.setValue(IdeBackgroundUtil.EDITOR_PROP, imageProp);
```

File: src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java (hunk 1 above)
Comment (drinchev) [reply src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java]: Yeah I checked the docs and we are forced to use a file here. I guess it's okay to choose the temp-file approach :)

File: src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        prop.setValue(IdeBackgroundUtil.EDITOR_PROP, null);
+        prop.setValue(IdeBackgroundUtil.FRAME_PROP, null);
+    }
+
+
+    private void createImage() {
+        if (bufferedImage == null) {
+            try {
+                ProjectLabelAWTRenderer renderer = new ProjectLabelAWTRenderer(preferences);
+                Dimension preferredImageRation = renderer.getPreferredImageRatio();
```
Comment (drinchev) [inline src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java]: I guess this makes the label a bit spooky.

So it will never have the same `height` but it will rather depend on the length of the label : 

A
A1
A1B

Do you think it will be better if we convert this to something like "20%" of the height of the screen ( so it also looks nice on a laptop size )?

File: src/main/java/com/drinchev/projectlabel/resources/ui/PluginConfiguration.java (hunk 2 above)
Comment (simschla) [reply src/main/java/com/drinchev/projectlabel/resources/ui/PluginConfiguration.java]: This is weird 🤔 
Is this a special theme or a special IntelliJ version on your end? The center icon seems identical to my screenshot, the corner icons are all the same (which is weird) and they all look different to the icons I see (which is also weird).
I see:
SCR-20230408-jdci

File: src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java (hunk 3 above)
Comment (simschla) [reply src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java]: thanks & sorry for that

File: src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java (hunk 4 above)
Comment (simschla) [reply src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java]: Yes, I'm also not entirely happy with the current solution.

I see the following options:

1.) **relative to the screen size**
This would mean we need to adapt the image whenever the screen size changes. (Display is plugged in/removed, Window changes Display)

2.) **relative to the window size**
Again, we would need to adapt it when the window changes size.

3.) **relative to the status bar widget font size**

4.) **Let the user select a font size in pixels**
Similar to the status bar widget, we let the user select a font size for the background image directly.

In all cases, it might be a good idea to somehow cap the maximum width of the image?

What's your take?

**Comment by simschla** [comment]
@drinchev What is your opinion on the project settings vs. application settings part?

File: src/main/java/com/drinchev/projectlabel/resources/ui/PluginConfiguration.java (hunk 2 above)
Comment (simschla) [reply src/main/java/com/drinchev/projectlabel/resources/ui/PluginConfiguration.java]: I've changed the code here so that we use custom icons now. Can you check if this fixes the problem for you?

Comment (drinchev) [reply src/main/java/com/drinchev/projectlabel/resources/ui/PluginConfiguration.java]: The icons now work properly. 

I'm also not sure what the reason is, but it seems that the `Run Plugin` IDE task might not start an IntelliJ instance that is the same as the ones we run.

image

**Comment by drinchev** [comment]
> @drinchev What is your opinion on the project settings vs. application settings part?

Sorry for the delayed replies. 

Yeah I think it will make more sense if we have the settings per project and global ( similar to the project label one ).

I guess we can show them in both places with some UI like this : 

image

This would be the only way for the user to actually change once the setting everywhere.

File: src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java (hunk 4 above)
Comment (drinchev) [reply src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java]: Yeah this gets quite ugly with two windows on the same machine.

image

I guess I would prefer to have **this relative to the window size**. 

I'm not sure it would be cool if the user chooses the font-size in pixels as I'm also moving sometimes from my desktop screen to my laptop screen and in one case the label is just too big. It doesn't make sense to change the font-size each time I move from desktop to laptop screen.

In the Status Bar we have the benefit that is always fixed height no matter which screen you are looking at it. If you think we can make the background label way-smaller ( something like 100px fixed height ) it would be okay for all screens I guess.

Comment (simschla) [reply src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java]: I've adapted the code to resize the label dynamically based on the window size. I kind of like it this way, give it a try and LMK what you think.

**Comment by simschla** [comment]
> Sorry for the delayed replies.
> 
> Yeah I think it will make more sense if we have the settings per project and global ( similar to the project label one ).
> 
> I guess we can show them in both places with some UI like this :
> 
> image
> 
> This would be the only way for the user to actually change once the setting everywhere.

I see where you're going with this, and it is quite elegant, I must say 👍 In addition to your screenshot, I think we should also apply this approach to the font settings of the status bar label (move "font family" and "font size" to one line and provide a single "Inherit" checkbox in the project prefs). That way, the UI would look cleaner and applies the same UX concept for the same situation. WDYT?

**Comment by drinchev** [comment]
> I think we should also apply this approach to the font settings of the status bar label (move "font family" and "font size" to one line and provide a single "Inherit" checkbox in the project prefs)

Sounds good :)

File: src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java (hunk 4 above)
Comment (drinchev) [reply src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java]: Looks way nicer. 

image

The only "bug" I see is the bottom margin is smaller than the right margin

Comment (simschla) [reply src/main/java/com/drinchev/projectlabel/resources/ui/ProjectLabelBackgroundImage.java]: Nice, I'll look into the margin issue and get back to you.

**Comment by simschla** [comment]
Update:
I've improved the rendering code so that the margin-error should now be gone.
I've adapted the settings dialog to what we discussed and streamlined its UI a bit.

SCR-20230418-kkdg
SCR-20230418-kjzc

Please check on your end again and let me know if you have any further adjustments you'd like me to make.

**Comment by drinchev** [comment]
@simschla Looks awesome. I will give it a try locally today or tomorrow and let you know in case I find anything else.

I hope we can publish this version over the weekend :)

**Comment by drinchev** [comment]
Hmm I don't know why but setting the preferences act wierd now 

https://user-images.githubusercontent.com/202919/233024121-1e7be900-e5f4-4d72-9fca-6a9e322f5492.mov

**Comment by simschla** [comment]
> Hmm I don't know why but setting the preferences act wierd now
> 
>  Screen.Recording.2023-04-19.at.10.56.59.mov

Hmmm, that doesn't look right 😉 I'll look into it, maybe I broke something when cleaning up the code.

**Comment by simschla** [comment]
> Hmmm, that doesn't look right 😉 I'll look into it, maybe I broke something when cleaning up the code.

Yeah, that was the case. -> should be fixed now.

I found another issue: when running on a HDPI display, something with padding is not correct. I'll look into it, but a second pair of eyes on that code part would be greatly appreciated.

**Comment by simschla** [comment]
> I found another issue: when running on a HDPI display, something with padding is not correct. I'll look into it, but a second pair of eyes on that code part would be greatly appreciated.

I was able to resolve that. You are welcome to give it another spin on your end.

File: src/main/java/com/drinchev/projectlabel/ProjectLabel.java [hunk 5]
```diff
@@ -1,21 +1,33 @@
...
 import com.intellij.openapi.wm.StatusBar;
 import com.intellij.openapi.wm.WindowManager;
+import org.jetbrains.annotations.NotNull;
+
+
+import static java.util.Objects.requireNonNull;
 
-@Service
-public final class ProjectLabelService {
+@Service(Level.PROJECT)
```
Comment (drinchev) [inline src/main/java/com/drinchev/projectlabel/ProjectLabel.java:16]: Do you have any specific reason for making this a Project Service, rather than being an Application one?

Being an Application Level Service allows us to change the global setting of the plugin and it will affect all open projects.

**Comment by drinchev** [comment]
I still can't really enable the background image globally.

image

Also it seems that the Label Font for the preferences is not having the same behaviour as before. 

I see a lot of changes in the code and is too hard to follow why the problem exists. 

Btw is there a way to get the icons from the IntelliJ package? https://jetbrains.design/intellij/resources/icons_list/

I'm not sure if we are actually able to include them in the source code, due to privacy concerns.

File: src/main/java/com/drinchev/projectlabel/ProjectLabel.java (hunk 5 above)
Comment (simschla) [reply src/main/java/com/drinchev/projectlabel/ProjectLabel.java:16]: I've made it a Project Service, since it only acts on project resources. I've just tried to change it to `APP` or without any specification locally - but that does not result in affecting all open projects. Is that something that worked before (meaning I broke it)?

**Comment by simschla** [comment]
> I still can't really enable the background image globally.

I'm sorry, I need to test this more thoroughly before wasting your time 🙇  Will try to look into it over the weekend.

 
> Also it seems that the Label Font for the preferences is not having the same behaviour as before.

Can you expand on that? I've changed the settings dialog so that when you "inherit", I try to update the project settings ui components, when the global components change.
 
> I see a lot of changes in the code and is too hard to follow why the problem exists.

The code that binds and updates the ui components depending on `inherit` and `enable` toggles is here:

https://github.com/drinchev/project-label/pull/19/files#diff-34bee4f400448d868a713a29d8480cd16caac92ed456b1f678e3e93d13cf91baR133-R179

You're welcome to change that if you can chase down the bug 👍 
 
> Btw is there a way to get the icons from the IntelliJ package? https://jetbrains.design/intellij/resources/icons_list/

I've [followed their guide](https://jetbrains.design/intellij/principles/icons/) to download, modify and include the icons based on their default icon set. (Reason to create our own icons was that the icons were different / wrong for Platform 2022.3.3 vs. 2023.1)

> I'm not sure if we are actually able to include them in the source code, due to privacy concerns.

I think that should be fine if I read their guide correctly.

**Comment by simschla** [comment]
I've tested the preferences page now and fixed two things I've seen (and you'd seen, too). 

The behavior when you have multiple project windows open and change the global setting in one of them, is identical to the current release version, just tested that on my end. But I think it would be cool if we could trigger a repaint in all open windows. I suggest we raise an issue for that and get at it in a separate PR, if you're ok with that?

**Comment by drinchev** [comment]
Hey @simschla,

Thanks a lot for your patience :) 

> But I think it would be cool if we could trigger a repaint in all open windows. I suggest we raise an issue for that and get at it in a separate PR, if you're ok with that?

Yeah I guess we can do that in another PR. 

I tested the plugin and it seems to be working properly now 🥳 . 

Let's merge this one :)

**Review by drinchev — APPROVED** [review]

**Comment by simschla** [comment]
> Thanks a lot for your patience :)

No worries, we all have a lot to do these days 😄 

> Yeah I guess we can do that in another PR.

OK, just opened an issue here, so we do not forget: #23 

> I tested the plugin and it seems to be working properly now 🥳 .
> 
> Let's merge this one :)

Awesome, thanks!

**Comment by simschla** [comment]
Published in `1.2.0`.
